src/Pos/WorkMode.hs

```diff
@@ -11,83 +12,119 @@ module Pos.WorkMode
+import           System.Wlog                 (CanLog, HasLoggerName (..), LoggerName)
+
+import           Pos.Block.BListener         (MonadBListener (..), onApplyBlocksStub,
+                                              onRollbackBlocksStub)
+import           Pos.Block.Core              (Block, BlockHeader)
+import           Pos.Block.Types             (Undo)
+import           Pos.Communication.PeerState (PeerStateCtx, PeerStateTag,
+                                              WithPeerState (..), clearPeerStateReal,
+                                              getAllStatesReal, getPeerStateReal)
+import           Pos.Context                 (NodeContext)
+import           Pos.Core                    (IsHeader)
+import           Pos.DB                      (MonadGState (..), NodeDBs)
+import           Pos.DB.Block                (MonadBlockDBWrite (..), dbGetBlockReal,
+                                              dbGetBlockReal', dbGetHeaderReal,
+                                              dbGetHeaderReal', dbGetUndoReal,
+                                              dbGetUndoReal', dbPutBlundReal)
+import           Pos.DB.Class                (MonadBlockDBGeneric (..), MonadDB (..),
+                                              MonadDBRead (..))
+import           Pos.DB.DB                   (gsAdoptedBVDataDB)
+import           Pos.DB.Redirect             (dbDeleteReal, dbGetReal, dbPutReal,
+                                              dbWriteBatchReal)
+import           Pos.Delegation.Class        (DelegationVar)
+import           Pos.Discovery               (MonadDiscovery (..), findPeersReal,
+                                              getPeersReal)
+import           Pos.Slotting.Class          (MonadSlots (..))
+import           Pos.Slotting.Impl.Sum       (currentTimeSlottingReal,
+                                              getCurrentSlotBlockingReal,
+                                              getCurrentSlotInaccurateReal,
+                                              getCurrentSlotReal)
+import           Pos.Slotting.MemState       (MonadSlotsData (..), getSlottingDataReal,
+                                              getSystemStartReal, putSlottingDataReal,
+                                              waitPenultEpochEqualsReal)
+import           Pos.Ssc.Class.Helpers       (SscHelpersClass)
+import           Pos.Ssc.Class.Types         (SscBlock)
+import           Pos.Ssc.Extra               (SscMemTag, SscState)
+import           Pos.Txp.MemState            (GenericTxpLocalData, TxpHolderTag)
+import           Pos.Util                    (Some (..))
+import           Pos.Util.JsonLog            (JsonLogConfig, jsonLogReal)
+import           Pos.Util.TimeWarp           (CanJsonLog (..))
+import           Pos.Util.Util               (PowerLift (..))
+import           Pos.WorkMode.Class          (MinWorkMode, TxpExtra_TMP, WorkMode)
+
+data RealModeContext ssc = RealModeContext
```

Comment:
This is a WIP branch. I haven't rewritten the context data types to tuples. Actually, it won't be that simple, because with tuples I won't be able to embed one context into another.

I think the best way to go is to write some TH that would generate those records together with `HasLens` instances.